hassio-google-drive-backup/backup/static/js/theme.js

```diff
@@ -0,0 +1,328 @@
+  }
+
+  saturate(change) {
+    const Pr = 0.299;
+    const Pg = 0.587;
+    const Pb = 0.114;
+    const P = Math.sqrt(
+      this.r * this.r * Pr + this.g * this.g * Pg + this.b * this.b * Pb
+    );
+
+    const R = this._clamp(P + (this.r - P) * change);
+    const G = this._clamp(P + (this.g - P) * change);
+    const B = this._clamp(P + (this.b - P) * change);
+    return new Color(R, G, B, this.a);
+  }
+
+  _clamp(value, min = 0, max = 255) {
+    if (value > max) {
+      return max;
+    }
+    if (value < min) {
+      return min;
+    }
+    return value;
+  }
+
+  toCss() {
+    return (
+      'rgba(' +
+      Math.floor(this.r) +
+      ',' +
+      Math.floor(this.g) +
+      ',' +
+      Math.floor(this.b) +
+      ',' +
+      this.a +
+      ')'
+    );
+  }
+
+  toHex() {
+    const componentToHex = (c) => {
+      var hex = c.toString(16);
+      return hex.length == 1 ? '0' + hex : hex;
+    };
+
+    return (
+      '#' +
+      componentToHex(Math.floor(this.r)) +
+      componentToHex(Math.floor(this.g)) +
+      componentToHex(Math.floor(this.b))
+    );
+  }
+
+  textColor() {
+    const luma = (0.299 * this.r + 0.587 * this.g + 0.114 * this.b) / 255;
+    return luma > 0.53 ? Color.black() : Color.white();
+  }
+
+  luminance() {
+    const rg =
+      Math.floor(this.r) <= 10
+        ? this.r / 3294.0
+        : Math.pow(this.r / 269.0 + 0.0513, 2.4);
+    const gg =
+      Math.floor(this.g) <= 10
+        ? this.g / 3294.0
+        : Math.pow(this.g / 269.0 + 0.0513, 2.4);
+    const bg =
+      Math.floor(this.b) <= 10
+        ? this.b / 3294.0
+        : Math.pow(this.b / 269.0 + 0.0513, 2.4);
+    return 0.2126 * rg + 0.7152 * gg + 0.0722 * bg;
+  }
+
+  contrast(other) {
+    let big = this.luminance();
+    let small = other.luminance();
+
+    if (big < small) {
+      const temp = big;
+      big = small;
+      small = temp;
+    }
+
+    return (big + 0.05) / (small + 0.05);
+  }
+
+  withAlpha(alpha) {
+    return new Color(this.r, this.g, this.b, alpha);
+  }
+}
+
+function setColors(background, accent) {
+  const text = background.textColor();
+  const accentText = accent.textColor();
+  const contrastThreshold = 4.5;
+  const contrast = background.contrast(accent);
+
+  let linkAccent = accent;
+  if (contrast < contrastThreshold) {
+    // do some adjustment to make the UI more readable if the contrast is really bad
+    const scale = 1 - (contrast - 1) / (contrastThreshold - 1);
+    linkAccent = linkAccent.tint(text, scale * 0.5);
+  }
+
+  const focus = accent.saturate(1.2);
+  const help = text.tint(background, 0.25);
+
+  const shadow1 = text.withAlpha(0.14);
+  const shadow2 = text.withAlpha(0.12);
+  const shadow3 = text.withAlpha(0.2);
+  const shadowBmc = background.withAlpha(0.2);
+  const bgShadow = `0 2px 2px 0 ${shadow1.toCss()}, 0 3px 1px -2px ${shadow2.toCss()}, 0 1px 5px 0 ${shadow3.toCss()}`;
+  const bgModal = background.tint(text, 0.02);
+
+  const styleSheet = {
```

Comment:
Here I create an object which contains all the CSS properties, and then I convert it to CSS text later. Imho it is easier to read like that.

Reply:
I agree, the way I was doing it was a bit of a mess.  Did the job, but left much to be desired.